Add methods to augment allowed headers and parameters in StrictHttpFi…

baezzys opened the pull request:
This pull request introduces new constants in the `StrictHttpFirewall` class that allow for the augmentation of the sets of allowable header names, header values, parameter names, and parameter values.

This closes #13639

jzheaux reviewed (changes requested):
Thanks, @baezzys! There is an issue with having both `set` and `add` that I didn't see earlier, which is that it assumes that the application wants to do `Predicate#and` and not `Predicate#or`. It doesn't give the developer much more power.

Also, I think that `addAllowHeaderValues` makes it seem like you are listing additional ways that headers would be allowed, which is the opposite of the implementation.

Instead, please let's introduce `public static` defaults for each, like so:

```java
public static final Predicate<String> ALLOWED_HEADER_NAMES = ...;

public static final Predicate<String> ALLOWED_HEADER_VALUES = ...;

public static final Predicate<String> ALLOWED_PARAMETER_NAMES = ...;

public static final Predicate<String> ALLOWED_PARAMETER_VALUES = ...;
```

So that an application can do:

```java
firewall.setAllowedHeaderValues(ALLOWED_HEADER_VALUES.and((value) -> !value.contains("\t")))
```

baezzys commented:
Thank you for the review, @jzheaux. However, since the setters are not static, would it be acceptable to change them to public instead of public static?

jzheaux commented:
I think they should be `static` so that you don't need an instance of the firewall to refer to them.

baezzys commented:
Thank you for the feedback, @jzheaux

I have updated the `StrictHttpFirewall` class to change the private Predicates for allowed header names, header values, parameter names, and parameter values to public static. Additionally, I have changed the corresponding setters.

Please review the changes and let me know if any further adjustments are needed.

jzheaux commented on web/src/main/java/org/springframework/security/web/firewall/StrictHttpFirewall.java:
```diff
@@ -134,13 +135,13 @@ public class StrictHttpFirewall implements HttpFirewall {
 
 	private static final Predicate<String> HEADER_VALUE_PREDICATE = (s) -> HEADER_VALUE_PATTERN.matcher(s).matches();
 
-	private Predicate<String> allowedHeaderNames = ASSIGNED_AND_NOT_ISO_CONTROL_PREDICATE;
```
Let's keep both. There is the public constant that folks can use to build expressions and then there is whatever in the internal setting. You'd want something like this:

```java
private Predicate<String> allowedHeaderNames = ALLOWED_HEADER_NAMES;

public static final Predicate<String> ALLOWED_HEADER_NAMES = ASSIGNED_AND_NOT_ISO_CONTROL_PREDICATE;
```

baezzys replied:
Thank you for your feedback, @jzheaux.

I have updated the code to reflect your suggestions. 

One question I have is regarding the potential confusion that might arise because ALLOWED_HEADER_NAMES is final and thus immutable. Even if the user changes the allowedHeaderNames predicate, the ALLOWED_HEADER_NAMES constant remains unchanged. Could this lead to any misunderstandings or misuses?

jzheaux replied:
This is a common standard in Java, so I don't think it will lead to wide confusion. The constant needs to be final because that is not the thing that is being changed. Instead, the setter is updating the value for `allowedHeaderNames`.

jzheaux commented on web/src/main/java/org/springframework/security/web/firewall/StrictHttpFirewall.java:
```diff
@@ -435,9 +436,9 @@ public void setAllowUrlEncodedLineSeparator(boolean allowUrlEncodedLineSeparator
 	 * @see Character#isDefined(int)
 	 */
-	public void setAllowedHeaderNames(Predicate<String> allowedHeaderNames) {
+	public static void setAllowedHeaderNames(Predicate<String> allowedHeaderNames) {
```
Please keep the methods non-static. We only need the public constants to be static.

baezzys replied:
Done.

jzheaux commented on web/src/main/java/org/springframework/security/web/firewall/StrictHttpFirewall.java:
```diff
@@ -435,9 +436,9 @@ public void setAllowUrlEncodedLineSeparator(boolean allowUrlEncodedLineSeparator
-	public void setAllowedHeaderNames(Predicate<String> allowedHeaderNames) {
+	public static void setAllowedHeaderNames(Predicate<String> allowedHeaderNames) {
 		Assert.notNull(allowedHeaderNames, "allowedHeaderNames cannot be null");
-		this.allowedHeaderNames = allowedHeaderNames;
```
Please keep this line as it was.

jzheaux commented on web/src/main/java/org/springframework/security/web/firewall/StrictHttpFirewall.java:
```diff
@@ -450,9 +451,9 @@ public void setAllowedHeaderNames(Predicate<String> allowedHeaderNames) {
+	public static void setAllowedHeaderValues(Predicate<String> allowedHeaderValues) {
 		Assert.notNull(allowedHeaderValues, "allowedHeaderValues cannot be null");
-		this.allowedHeaderValues = allowedHeaderValues;
+		ALLOWED_HEADER_VALUES = allowedHeaderValues;
```
Please keep this line as it was.

jzheaux commented on web/src/main/java/org/springframework/security/web/firewall/StrictHttpFirewall.java:
```diff
@@ -463,9 +464,9 @@ public void setAllowedHeaderValues(Predicate<String> allowedHeaderValues) {
-	public void setAllowedParameterNames(Predicate<String> allowedParameterNames) {
+	public static void setAllowedParameterNames(Predicate<String> allowedParameterNames) {
 		Assert.notNull(allowedParameterNames, "allowedParameterNames cannot be null");
-		this.allowedParameterNames = allowedParameterNames;
```
Please keep this line as it was

baezzys replied:
Done.

jzheaux commented on web/src/main/java/org/springframework/security/web/firewall/StrictHttpFirewall.java:
```diff
@@ -476,9 +477,9 @@ public void setAllowedParameterNames(Predicate<String> allowedParameterNames) {
-	public void setAllowedParameterValues(Predicate<String> allowedParameterValues) {
+	public static void setAllowedParameterValues(Predicate<String> allowedParameterValues) {
 		Assert.notNull(allowedParameterValues, "allowedParameterValues cannot be null");
-		this.allowedParameterValues = allowedParameterValues;
```
Please keep this line as it was

jzheaux commented on web/src/main/java/org/springframework/security/web/firewall/StrictHttpFirewall.java:
```diff
@@ -826,28 +827,28 @@ public String[] getParameterValues(String name) {
 
 		private void validateAllowedHeaderName(String headerNames) {
-			if (!StrictHttpFirewall.this.allowedHeaderNames.test(headerNames)) {
+			if (!StrictHttpFirewall.this.ALLOWED_HEADER_NAMES.test(headerNames)) {
```
Please leave this line as it was.

jzheaux commented on web/src/main/java/org/springframework/security/web/firewall/StrictHttpFirewall.java:
```diff
@@ -826,28 +827,28 @@ public String[] getParameterValues(String name) {
 
 		private void validateAllowedHeaderValue(String name, String value) {
-			if (!StrictHttpFirewall.this.allowedHeaderValues.test(value)) {
+			if (!StrictHttpFirewall.this.ALLOWED_HEADER_VALUES.test(value)) {
```
Please leave this line as it was.

jzheaux commented on web/src/main/java/org/springframework/security/web/firewall/StrictHttpFirewall.java:
```diff
@@ -826,28 +827,28 @@ public String[] getParameterValues(String name) {
 
 		private void validateAllowedParameterName(String name) {
-			if (!StrictHttpFirewall.this.allowedParameterNames.test(name)) {
+			if (!StrictHttpFirewall.this.ALLOWED_PARAMETER_NAMES.test(name)) {
```
Please leave this line as it was.

baezzys replied:
Done.

jzheaux commented on web/src/main/java/org/springframework/security/web/firewall/StrictHttpFirewall.java:
```diff
@@ -826,28 +827,28 @@ public String[] getParameterValues(String name) {
 
 		private void validateAllowedParameterValue(String name, String value) {
-			if (!StrictHttpFirewall.this.allowedParameterValues.test(value)) {
+			if (!StrictHttpFirewall.this.ALLOWED_PARAMETER_VALUES.test(value)) {
```
Please leave this line as it was.

baezzys replied:
Done.

jzheaux commented on web/src/test/java/org/springframework/security/web/firewall/StrictHttpFirewallTests.java:
```diff
@@ -41,6 +44,19 @@ public class StrictHttpFirewallTests {
+
+	@BeforeEach
+	public void initializeFirewall() {
+		StrictHttpFirewall.ALLOWED_HEADER_NAMES = (s) -> ASSIGNED_AND_NOT_ISO_CONTROL_PATTERN.matcher(s).matches();
```
I think these will already be set by the object itself. They won't need to be reset if you follow the earlier suggestions.

baezzys replied:
Done.

jzheaux commented on web/src/test/java/org/springframework/security/web/firewall/StrictHttpFirewallTests.java:
```diff
@@ -723,6 +739,18 @@ public void getFirewalledRequestGetHeaderWhenNotAllowedHeaderNameThenException()
+	public void getFirewalledRequestGetHeaderWhenSetNotAllowedHeaderNamesThenException() {
+		this.firewall.setAllowedHeaderNames((name) -> !name.equals("bad name"));
+		this.firewall
+			.setAllowedHeaderNames(this.firewall.ALLOWED_HEADER_NAMES.and((name) -> !name.equals("worst name")));
```
After the above changes, this will change to something closer to:

```java
this.firewall.setAllowedHeaderNames(
    StrictHttpFirewall.ALLOWED_HEADER_NAMES.and((name) -> !name.equals("worst name"));
```

baezzys replied:
Done.

jzheaux commented on web/src/test/java/org/springframework/security/web/firewall/StrictHttpFirewallTests.java:
```diff
@@ -731,6 +759,21 @@ public void getFirewalledRequestGetHeaderWhenNotAllowedHeaderValueThenException(
+		this.request.addHeader("good name", "bad value");
+		this.request.addHeader("best name", "worst value");
+
+		this.firewall.setAllowedHeaderValues((value) -> !value.equals("bad value"));
```
Since the second call overrides the first, you can leave this line out and it won't change the result of the test.

baezzys replied:
Thank you for the review.

I have updated the code to retain only the first call.

jzheaux reviewed (changes requested):
Thanks, @baezzys! I've left some feedback inline.

baezzys commented:
Hi @jzheaux, Sorry for the late response. I've made the changes based on your feedback. PTAL. Thanks!

jzheaux commented:
Thank you, @baezzys, for the PR! This is now merged into `main`.
